pkg/gardenlet/controller/federatedseed/shootstate/common_sync.go

```diff
@@ -0,0 +1,156 @@
+		return
+	}
+	queue.Add(key)
+}
+
+func (s *SyncController) createShootStateSyncReconcileFunc(ctx context.Context, kind string, obj runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		obj = obj.DeepCopyObject()
+		err := s.seedClient.Client().Get(ctx, req.NamespacedName, obj)
+		if apierrors.IsNotFound(err) {
+			s.log.Debugf("Skipping ShootState sync because resoruce with kind %s is missing", kind)
+			return reconcile.Result{}, nil
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		extensionObject, err := apiextensions.Accessor(obj)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		state := extensionObject.GetExtensionStatus().GetState()
+		if state == nil {
+			return reconcile.Result{}, fmt.Errorf("invalid state. Extension state for resource with kind %s in namespace %s was changed to nil before it could be synced to ShootState", kind, req.Namespace)
+		}
+
+		name := extensionObject.GetName()
+		purpose := getPurpose(obj)
+		projectNamespace, shootName := fromNamespaceName(req.Namespace)
+		shootState := &gardencorev1alpha1.ShootState{ObjectMeta: metav1.ObjectMeta{Name: shootName, Namespace: projectNamespace}}
+		if _, err := controllerutil.CreateOrUpdate(ctx, s.k8sGardenClient.Client(), shootState, func() error {
+			return updateShootStateExtensionState(state, shootState, kind, name, purpose)
+		}); err != nil {
+			message := fmt.Sprintf("Shoot %s extension state was NOT succcessfuly synced", kind)
+			s.log.Error(message)
+			s.recorder.Event(shootState, corev1.EventTypeNormal, "ScheduledNextSync", message)
+			return reconcile.Result{}, err
+		}
+
+		message := fmt.Sprintf("Shoot %s extension state was succcessfuly synced", kind)
+		s.log.Info(message)
+		s.recorder.Event(shootState, corev1.EventTypeNormal, "ScheduledNextSync", message)
+		return reconcile.Result{Requeue: false}, nil
+	}
+}
+
+func updateShootStateExtensionState(extensionState *runtime.RawExtension, shootState *gardencorev1alpha1.ShootState, kind string, name string, purpose *string) error {
+	currentExtensionState := gardencorev1alpha1helper.GetExtensionResourceState(shootState.Spec.Extensions, kind, &name, purpose)
+	if currentExtensionState == nil {
+		shootState.Spec.Extensions = append(shootState.Spec.Extensions, gardencorev1alpha1.ExtensionResourceState{
+			Kind:    kind,
+			Name:    &name,
+			Purpose: purpose,
+			State:   gardencorev1alpha1.ProviderConfig{RawExtension: *extensionState},
+		})
+	} else {
+		currentExtensionState.State = gardencorev1alpha1.ProviderConfig{RawExtension: *extensionState}
+	}
+	return nil
+}
+
+func getPurpose(obj runtime.Object) *string {
+	switch v := obj.(type) {
+	case *extensionsv1alpha1.ControlPlane:
+		return (*string)(v.Spec.Purpose)
+	case *extensionsv1alpha1.OperatingSystemConfig:
+		return (*string)(&v.Spec.Purpose)
+	default:
+		return nil
+	}
+}
+
+func fromNamespaceName(namespace string) (projectNamespace, shootName string) {
```

Comment:
Can you move this function below https://github.com/gardener/gardener/blob/master/pkg/operation/shoot/shoot.go#L219-L231, export it and call it `UnfoldTechnicalID` (or something similar) + implement a unit test?